lib/spack/spack/environment.py

```diff
@@ -131,73 +131,105 @@ def deactivate():
+            return read(env)
+
+    else:
+        # if env was specified, see if it is a dirctory otherwise, look
+        # at env_dir (env and env_dir are mutually exclusive)
+        env = getattr(args, 'env_dir', None)
+
+        # if no argument, look for a manifest file
+        if not env:
+            if os.path.exists(manifest_name):
+                env = os.getcwd()
 
-    environment = disambiguate(env)
+            # if no env, env_dir, or manifest try the environment
+            if not env:
+                env = os.environ.get(spack_env_var)
 
-    if not environment:
-        tty.die('no such environment: %s' % env)
-    return environment
+                # nothing was set; there's no active environment
+                if not env:
+                    return None
+
+    # if we get here, env isn't the name of a spack environment; it has
+    # to be a path to an environment, or there is something wrong.
+    if is_env_dir(env):
+        return Environment(env)
+
+    raise SpackEnvironmentError('no environment in %s' % env)
+
+
+def get_env(args, cmd_name, required=True):
+    """Used by commands to get the active environment.
 
+    This first checks for an ``env`` argument, then looks at the
+    ``active`` environment.  We check args first because Spack's
+    subcommand arguments are parsed *after* the ``-e`` and ``-D``
+    arguments to ``spack``.  So there may be an ``env`` argument that is
+    *not* the active environment, and we give it precedence.
 
-def disambiguate(env, env_dir=None):
-    """Used to determine whether an environment is named or a directory."""
+    This is used by a number of commands for determining whether there is
+    an active environment.
+
+    If an environment is not found *and* is required, print an error
+    message that says the calling command *needs* an active environment.
+
+    Arguments:
+        args (Namespace): argparse namespace wtih command arguments
+        cmd_name (str): name of calling command
+        required (bool): if ``False``, return ``None`` if no environment
+            is found instead of raising an exception.
+
+    Returns:
+        (Environment): if there is an arg or active environment
+    """
+    # try argument first
+    env = getattr(args, 'env', None)
     if env:
         if exists(env):
-            # treat env as a name
             return read(env)
-        env_dir = env
-
-    if not env_dir:
-        env_dir = os.environ.get(spack_env_var)
-        if not env_dir:
-            return None
-
-    if os.path.isdir(env_dir):
-        if is_env_dir(env_dir):
-            return Environment(env_dir)
+        elif is_env_dir(env):
+            return Environment(env)
         else:
-            raise SpackEnvironmentError('no environment in %s' % env_dir)
-
-    return None
+            raise SpackEnvironmentError('no environment in %s' % env)
+
+    # try the active environment. This is set by find_environment() (above)
+    if active:
```

Comment:
@tgamblin does this need to designate `active` as a global before checking it?

Also, can the global have some naming convention that distinguishes it from local variables? All caps? Adding "global" as a prefix (e.g. `active_global`)?